pulsar/producer_test.go

```diff
@@ -2474,3 +2476,33 @@ func TestDisableReplication(t *testing.T) {
 	assert.NoError(t, err)
 	assert.Equal(t, []string{"__local__"}, msgMetadata.GetReplicateTo())
 }
+
+func TestProducerWithMaxConnectionsPerBroker(t *testing.T) {
```

Comment:
The change looks good to me. But how this test case reproduce the issues?

Reply:
```go
func TestSimpleProducer(t *testing.T) {
	client, err := NewClient(ClientOptions{
		URL: serviceURL,
		MaxConnectionsPerBroker: 2,
	})
	assert.NoError(t, err)
	defer client.Close()

	producer, err := client.CreateProducer(ProducerOptions{
		Topic: newTopicName(),
	})

	assert.NoError(t, err)
	assert.NotNil(t, producer)
	defer producer.Close()

	for i := 0; i < 10; i++ {
		ID, err := producer.Send(context.Background(), &ProducerMessage{
			Payload: []byte("hello"),
		})

		assert.NoError(t, err)
		assert.NotNil(t, ID)
	}

	
}
```
Under this test scenario, the issue can be triggered as long as the `MaxConnectionsPerBroker` is greater than 1.

Because the connection obtained here:
https://github.com/apache/pulsar-client-go/blob/50dce7ea7c2d7aff64c65ccbc2ef696a25cc9afc/pulsar/producer_partition.go#L285-L289

is not the same as the connection obtained there:
https://github.com/apache/pulsar-client-go/blob/50dce7ea7c2d7aff64c65ccbc2ef696a25cc9afc/pulsar/producer_partition.go#L295


This will lead to the following exception:

```log
time="2024-07-12T09:52:25+08:00" level=error msg="Failed to create producer at send PRODUCER request" error="server error: ProducerBusy: org.apache.pulsar.broker.service.BrokerServiceException$NamingException: Producer with name 'standalone-6-9' is already connected to topic 'persistent://public/default/my-topic-348398000'" producerID=1 producer_name=standalone-6-9 topic="persistent://public/default/my-topic-348398000"
time="2024-07-12T09:52:25+08:00" level=error msg="Failed to create producer at reconnect" error="server error: ProducerBusy: org.apache.pulsar.broker.service.BrokerServiceException$NamingException: Producer with name 'standalone-6-9' is already connected to topic 'persistent://public/default/my-topic-348398000'" producerID=1 producer_name=standalone-6-9 topic="persistent://public/default/my-topic-348398000"
```

Reply:
Thanks. I get it now.